src/repr/src/adt/range.rs

```diff
@@ -0,0 +1,365 @@
+    pub inclusive: bool,
+    /// None value represents an infinite bound.
+    pub bound: Option<DatumNested<'a>>,
+}
+
+impl<'a, const UPPER: bool> Display for RangeBound<'a, UPPER> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.bound {
+            None => Ok(()),
+            Some(bound) => Display::fmt(&bound.datum(), f),
+        }
+    }
+}
+
+impl<'a, const UPPER: bool> Ord for RangeBound<'a, UPPER> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let ordering = match self.bound.cmp(&other.bound) {
+            Ordering::Equal => {
+                if self.inclusive == other.inclusive {
+                    Ordering::Equal
+                } else if self.inclusive {
+                    Ordering::Less
+                } else {
+                    Ordering::Greater
+                }
+            }
+            o => o,
+        };
+        if UPPER {
+            ordering.reverse()
+        } else {
+            ordering
+        }
+    }
+}
+
+impl<'a, const UPPER: bool> PartialOrd for RangeBound<'a, UPPER> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A `RangeBound` that sorts correctly for use as a lower bound.
+pub type RangeLowerBound<'a> = RangeBound<'a, false>;
+
+/// A `RangeBound` that sorts correctly for use as an upper bound.
+pub type RangeUpperBound<'a> = RangeBound<'a, true>;
+
+impl<'a, const UPPER: bool> RangeBound<'a, UPPER> {
+    /// Determines where `elem` lies in relation to the range bound.
+    ///
+    /// # Panics
+    /// - If `self.bound.datum()` is not convertible to `T`.
+    fn elem_cmp<T: RangeOps<'a>>(&self, elem: &T) -> Ordering {
+        match self.bound.map(|bound| {
+            <T>::try_from(bound.datum())
+                .unwrap_or_else(|_| panic!("cannot take {} to {}", bound.datum(), type_name::<T>()))
+        }) {
+            None if UPPER => Ordering::Greater,
+            None => Ordering::Less,
+            Some(bound) => bound.cmp(elem),
+        }
+    }
+
+    /// Does `elem` satisfy this bound?
+    fn satisfied_by<T: RangeOps<'a>>(&self, elem: &T) -> bool {
+        match self.elem_cmp(elem) {
+            // Inclusive always satisfied with equality, regardless of upper or
+            // lower.
+            Ordering::Equal => self.inclusive,
+            // Upper satisfied with values less than itself
+            Ordering::Greater => UPPER,
+            // Lower satisfied with values greater than itself
+            Ordering::Less => !UPPER,
+        }
+    }
+}
+
+/// Describes the value passed in via `RangeBoundDesc`s.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, PartialOrd)]
+pub enum RangeBoundDescValue<D> {
+    Finite { value: D },
+    Infinite,
+}
+
+impl<'a> From<Datum<'a>> for RangeBoundDescValue<Datum<'a>> {
+    /// Treats `Datum::Null` as an infinite bound (appropriate for PG
+    /// semantics), and all other `Datum`s as finite values.
```

Comment:
This feels a _little_ sketchy to me. I know it's what PostgreSQL does, but I'd almost prefer we do it in the one spot that PostgreSQL does it (range constructors, I think?) and nowhere else? But, idk, it's hard to know without seeing the rest of the code. No strong feels for the moment!

Reply:
Yeah--I think this API will shift as we build out SQL support for range types, but want to merge as is for sake convenience when building out the reclock migration.